README
ctt - time tracking for geeks

Type "make" to generate the manpage.


changelog
Changelog
---------

	* Changes are always commented with their author in (braces)
	* Exception: No braces means author == Nico Schottelius
	* Ignore, warn and continue in case of invalid delta (Darko Poljak)

1.1: 2017-02-16
	* Ignore non matching patterns for report command (Darko Poljak)
	* Added -s, --summary option (Darko Poljak)
	* No args error (Darko Poljak)
	* Report project name as file path basename (Darko Poljak)
	* Report globbing (Darko Poljak)

1.0: 2014-07-01
	* Added installer (Oz Nahum)
	* Bugfix in listprojects (Oz Nahum)
	* Allow ctt to run from checkout and installation path
	* Correct license indication in setup.py (Davide Riccardo Caliendo)
	* Added license and manifest files (Davide Riccardo Caliendo)
	* Added bash completion script (Davide Riccardo Caliendo)
	* Report an error on invalid time entries


0.9: 2013-04-29
	* Renamed -s/-e to --sd / --ed for tracking as well
	* Added manpage as documentation

0.8: 2013-04-27
	* Added -e / --regexp support for filtering entries on report
	* Added -i / --ignore-case support
	* Renamed -s/-e to --sd / --ed (to allow -e for expression)

0.7:
	* Added -f / --format support for reporting


lib/ctt/test/__init__.py
# -*- coding: utf-8 -*-
#
# 2016 Darko Poljak (darko.poljak at gmail.com))
#
# This file is part of ctt.
#
# cdist is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# cdist is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ctt. If not, see <http://www.gnu.org/licenses/>.
#
#

import os
import unittest

fixtures_dir = os.path.abspath(os.path.join(os.path.dirname(__file__),
                                            "fixtures"))


class CttTestCase(unittest.TestCase):
    def setUp(self):
        os.environ['HOME'] = fixtures_dir


lib/ctt/test/__main__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# 2016 Darko Poljak (darko.poljak at gmail.com)
#
# This file is part of ctt.
#
# cdist is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# cdist is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ctt. If not, see <http://www.gnu.org/licenses/>.
#
#

import imp
import os
import os.path
import sys
import unittest

base_dir = os.path.dirname(os.path.realpath(__file__))

test_modules = []
for possible_test in os.listdir(base_dir):
    if possible_test.startswith('test_'):
        mod_path = os.path.join(base_dir, possible_test)
        if os.path.isfile(mod_path):
            module = os.path.splitext(os.path.basename(possible_test))[0]
            test_modules.append(module)

suites = []
for test_module in test_modules:
    module_parameters = imp.find_module(test_module, [base_dir])
    module = imp.load_module("ctt.test." + test_module, *module_parameters)

    suite = unittest.defaultTestLoader.loadTestsFromModule(module)
    suites.append(suite)

all_suites = unittest.TestSuite(suites)
unittest.TextTestRunner(verbosity=2).run(all_suites)


lib/ctt/test/test_report.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# 2016 Darko Poljak (darko.poljak at gmail.com)
#
# This file is part of ctt.
#
# ctt is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ctt is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ctt. If not, see <http://www.gnu.org/licenses/>.
#
#

import unittest
import ctt
import ctt.test
import ctt.report as report
import sys
from io import StringIO
import datetime
import collections


class ReportTestCase(ctt.test.CttTestCase):
    def setUp(self):
        super(ReportTestCase, self).setUp()
        self.sys_stdout = sys.stdout
        out = StringIO()
        sys.stdout = out
        self.maxDiff = None

    def _get_output(self):
        sys.stdout.flush()
        output = sys.stdout.getvalue().strip()
        return output

    def tearDown(self):
        sys.stdout = self.sys_stdout
        super(ReportTestCase, self).tearDown()

    def test_print_report_time_entries(self):
        report_data = {
            '2016-04-07-0826': [
                {
                    'start_datetime': '2016-04-07-0826',
                    'end_datetime': '2016-04-07-2359',
                    'comment': 'foo1',
                    'delta': '6',
                    'delta_seconds': '6',
                    'delta_minutes': '0',
                },
            ],
        }
        report.Report.print_report_time_entries(report_data, ctt.REPORTFORMAT,
                                                False)
        output = self._get_output()
        expected_output = "2016-04-07-0826 (6): foo1"
        self.assertEqual(output, expected_output)

    def test_print_report_time_entries_summary(self):
        report_data = {
            '2016-04-07-0826': [
                {
                    'start_datetime': '2016-04-07-0826',
                    'end_datetime': '2016-04-07-2359',
                    'comment': 'foo1',
                    'delta': '6',
                    'delta_seconds': '6',
                    'delta_minutes': '0',
                },
            ],
            '2016-04-07-0926': [
                {
                    'start_datetime': '2016-04-07-0926',
                    'end_datetime': '2016-04-07-2359',
                    'comment': 'foo12',
                    'delta': '10',
                    'delta_seconds': '10',
                    'delta_minutes': '0',
                },
            ],
        }
        report.Report.print_report_time_entries(report_data, ctt.REPORTFORMAT,
                                                True)
        output = self._get_output()
        expected_output = ("2016-04-07-0826 (6): foo1\n"
                           "2016-04-07-0926 (10): foo12")
        self.assertEqual(output, expected_output)

    def test_print_reports(self):
        reports = collections.OrderedDict()
        for project in ('foo1', 'foo2'):
            rep = report.Report(project, ('2016-04-07', ), ('2016-04-08', ),
                                ctt.REPORTFORMAT, None, None)
            report_data = rep.report()
            reports[project] = (rep, report_data)
        expected_output = (
            "Report for foo1 between 2016-04-07 00:00:00 and "
            "2016-04-08 23:59:59\n"
            "2016-04-07-0826 (0:00:06): foo1\n"
            "2016-04-08-1200 (1:23:20): foo1 12\n"
            "Report for foo2 between 2016-04-07 00:00:00 and "
            "2016-04-08 23:59:59\n"
            "2016-04-07-0810 (0:00:10): foo2"
        )
        rep.print_reports(reports, ctt.REPORTFORMAT, summary=False)
        output = self._get_output()
        self.assertEqual(output, expected_output)

    # Summary should not print time entries
    # def test_print_reports_summary(self):
    #     reports = collections.OrderedDict()
    #     for project in ('foo1', 'foo2'):
    #         rep = report.Report(project, ('2016-04-07',), ('2016-04-08',),
    #                 ctt.REPORTFORMAT, None, None)
    #         report_data = rep.report()
    #         reports[project] = (rep, report_data)
    #     expected_output = (
    #         "2016-04-07-0810 (0:00:10): foo2\n"
    #         "2016-04-07-0826 (0:00:06): foo1\n"
    #         "2016-04-08-1200 (1:23:20): foo1 12"
    #     )
    #     rep.print_reports(reports, ctt.REPORTFORMAT, summary=True)
    #     output = self._get_output()
    #     self.assertEqual(output, expected_output)

    def test__init_date(self):
        rep = report.Report('foo1', ('2016-04-07', ), ('2016-04-07', ),
                            ctt.REPORTFORMAT, None, None)
        expected_start_date = datetime.datetime(2016, 4, 7)
        expected_end_date = datetime.datetime(2016, 4, 7, 23, 59, 59)
        self.assertEqual(rep.start_date, expected_start_date)
        self.assertEqual(rep.end_date, expected_end_date)

    @unittest.expectedFailure
    def test__init_date_fail(self):
        report.Report('foo1', ('2016-04-08', ), ('2016-04-07', ),
                      ctt.REPORTFORMAT, None, None)

    def test__init_date_defaults(self):
        rep = report.Report('foo1', None, None,
                            ctt.REPORTFORMAT, None, None)
        now = datetime.datetime.now()
        expected_start_date = now.replace(day=1, hour=0, minute=0, second=0,
                                          microsecond=0)
        next_month = expected_start_date.replace(day=28) + datetime.timedelta(
            days=4)
        first_day_next_month = next_month.replace(day=1)
        expected_end_date = first_day_next_month - datetime.timedelta(
            seconds=1)
        self.assertEqual(rep.start_date, expected_start_date)
        self.assertEqual(rep.end_date, expected_end_date)

    @unittest.expectedFailure
    def test__init_report_db_fail(self):
        report.Report('unexisting', ('2016-04-07',), ('2016-04-07',),
                      ctt.REPORTFORMAT, None, None)

    def test__init_report_db(self):
        rep = report.Report('foo1', ('2016-04-07',), ('2016-04-07',),
                            ctt.REPORTFORMAT, None, None)
        expected_db = {
            '2016-04-07-0826': {
                'comment': 'foo1',
                'delta': '6.248274'
            },
        }
        self.assertEqual(rep._report_db, expected_db)

    def test_header(self):
        rep = report.Report('foo1', ('2016-04-07',), ('2016-04-07',),
                            ctt.REPORTFORMAT, None, None)
        rep.header()
        output = self._get_output()
        self.assertEqual(output, ("Report for foo1 between 2016-04-07 00:00:00"
                                  " and 2016-04-07 23:59:59"))

    def test_summary(self):
        report.Report.summary(10)
        output = self._get_output()
        self.assertEqual(output, "Total time tracked: 0h 0m 10s.")

    def test_total_time(self):
        rep = report.Report('foo1', ('2016-04-07',), ('2016-04-07',),
                            ctt.REPORTFORMAT, None, None)
        total_time = rep.total_time
        expected_total_time = 6.248274
        self.assertEqual(total_time, expected_total_time)

    def test_report(self):
        rep = report.Report('foo1', ('2016-04-07',), ('2016-04-08',),
                            ctt.REPORTFORMAT, None, None)
        expected_entries = {
            '2016-04-07-0826': [
                {
                    'start_datetime': '2016-04-07-0826',
                    'end_datetime': '2016-04-07-0826',
                    'comment': 'foo1',
                    'delta': datetime.timedelta(seconds=6),
                    'delta_seconds': 6,
                    'delta_minutes': 0,
                },
            ],
            '2016-04-08-1200': [
                {
                    'start_datetime': '2016-04-08-1200',
                    'end_datetime': '2016-04-08-1323',
                    'comment': 'foo1 12',
                    'delta': datetime.timedelta(seconds=5000),
                    'delta_seconds': 5000,
                    'delta_minutes': 83,
                },
            ],
        }
        entries = rep.report()
        self.assertEqual(entries, expected_entries)


if __name__ == '__main__':
    unittest.main()


ctt.py
#!/usr/bin/env python3

import argparse
import calendar
import datetime

#import signal

import locale
import logging
import time

import os
import os.path
import sys

log = logging.getLogger(__name__)

VERSION = "0.1"

# Our output format
def user_datetime(when):
    """Print time for the user"""
    return when.ctime()

class Tracker:
    def __init__(self, project):
        self.project = project
        self.tracked_time = False

        self._init_home()


    def _init_home(self):
        # Create output directory
        home = os.environ['HOME']
        self.ctt_dir = os.path.join(home, ".ctt")
        self.project_dir = os.path.join(self.ctt_dir, self.project)


    # Track time and return information from tracking
    def track_time(self):
        self.start = datetime.datetime.now()

        try:
            # Wait for the exception to pop up - FIXME: find better method

            # Using input, Ctrl-C is displayed as ^C on the screen - ugly
            #input()

            # Sleep 42 years - should be longer than anybody trying to track time
            time.sleep(86400 * 365 * 42)

        except KeyboardInterrupt:
            pass

        self.stop = datetime.datetime.now()

        self.tracked_time = True

    def write_time(self):
        if not self.tracked_time:
            return

        start_seconds =  self.start.strftime("%s")
        stop_seconds =  self.stop.strftime("%s")
        delta_seconds = self.delta()

        time_dir = os.path.join(self.project_dir, start_seconds)
        os.makedirs(time_dir, mode=0o700, exist_ok=True)
        filename = os.path.join(time_dir, "delta")

        with open(filename, "w") as fd:
            fd.write("%s\n" % delta_seconds)

    def delta(self, in_seconds=True):
        if self.tracked_time:
            delta = self.stop - self.start
        else:
            delta = datetime.timedelta()

        if in_seconds:
            delta = delta.total_seconds()

        return delta


class Report(object):
    """Create a report on tracked time"""

    def __init__(self, args):
        pass

    @staticmethod
    def default_dates():
        """Return default start and end of of time
        start: first of last month
        end: last of last month
        """

        now = datetime.datetime.now()
        first_day = now.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
        end_date = first_day - datetime.timedelta(days=1)
        start_date = end_date.replace(day=1)

        return (start_date, end_date)


# Setup locale for calendar printing
# Setup locale to get Timezone information?
#print(locale.getlocale())

# Record project
# Record tags

def cmd_track(args):
    """Command line handler for time tracking"""
    tracker = Tracker(args.project[0])
    tracker.track_time()
    tracker.write_time()
    print(tracker.delta())

def cmd_report(args):
    """Command line handler for time reporting"""
    #report = Report(args.project[0])
    print(Report.default_dates())

def parse_argv(argv):
    parser = {}
    parser['main'] = argparse.ArgumentParser(description='ctt ' + VERSION)
    parser['sub'] = parser['main'].add_subparsers(title="Commands")

    parser['track'] = parser['sub'].add_parser('track')
    parser['track'].set_defaults(func=cmd_track)
    parser['track'].add_argument("project", help="Project to track time for", nargs=1)

    parser['report'] = parser['sub'].add_parser('report')
    parser['report'].set_defaults(func=cmd_report)
    parser['report'].add_argument("project", help="Project to report time for", nargs=1)
    parser['report'].add_argument("-s", "--start", help="Start datetime (first of last month)", nargs=1)
    parser['report'].add_argument("-e", "--end", help="End datetime (last of last month)", nargs=1)

    #parser['track'].add_argument("-t", "--tag", help="Add tags",
    #    action="store_true")

    args = parser['main'].parse_args()
    print(args)

    args.func(args)

if __name__ == "__main__":
    parse_argv(sys.argv[1:])
    sys.exit(0)

# Setup signal handler
# Start tracking
# Save stuff to our home directory

# Create datetime from userinput
# Wed Aug  1 23:35:53 2012


lib/ctt/tracker.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# 2012 Nico Schottelius (nico-ctt at schottelius.org)
#
# This file is part of ctt.
#
# ctt is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ctt is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ctt. If not, see <http://www.gnu.org/licenses/>.
#
#

import datetime
import logging
import os
import os.path

import ctt

log = logging.getLogger(__name__)


class Tracker:
    def __init__(self, project, start_datetime=None, end_datetime=None,
                 comment=True):
        self.project = project
        self.project_dir = ctt.project_dir(project)

        self._tracked_time = False
        self.comment = None

        # Setup default values
        try:
            if start_datetime:
                self.start_datetime = datetime.datetime.strptime(
                    start_datetime[0], ctt.DATETIMEFORMAT)
            else:
                self.start_datetime = None

            if end_datetime:
                self.end_datetime = datetime.datetime.strptime(
                   end_datetime[0], ctt.DATETIMEFORMAT)
            else:
                self.end_datetime = None
        except ValueError as e:
            raise ctt.Error(e)

        if self.start_datetime and self.end_datetime:
            self._tracked_time = True

    @classmethod
    def commandline(cls, args):
        tracker = cls(args.project[0], args.start, args.end, args.comment)
        tracker.track_time()

        if args.comment:
            tracker.record_comment()

        tracker.write_time()
        log.info(tracker.delta())

    def record_comment(self):
        """Record a comment for tracked data"""
        self.comment = input("Comment: ")

    # Track time and return information from tracking
    def track_time(self):
        """Track time, if necessary"""

        # Do not track again
        if self._tracked_time:
            return

        # If not given by the user
        if not self.start_datetime:
            self.start_datetime = datetime.datetime.now()

        try:
            # Wait for the exception to pop up
            input()
        except KeyboardInterrupt:
            pass

        self.end_datetime = datetime.datetime.now()

        self._tracked_time = True

    def write_time(self):
        if not self._tracked_time:
            return

        if self.start_datetime >= self.end_datetime:
            raise ctt.Error("End date must be after start date! (%s > %s)!" %
                            (self.start_datetime, self.end_datetime))

        subdirname = self.start_datetime.strftime(ctt.DISKFORMAT)
        time_dir = os.path.join(self.project_dir, subdirname)

        if os.path.exists(time_dir):
            raise ctt.Error(
                "Already tracked time at this beginning for this project")

        os.makedirs(time_dir, mode=0o700)

        filename = os.path.join(time_dir, ctt.FILE_DELTA)
        with open(filename, "w") as fd:
            fd.write("%s\n" % self.delta())

        if self.comment:
            filename = os.path.join(time_dir, ctt.FILE_COMMENT)
            with open(filename, "w") as fd:
                fd.write("%s\n" % self.comment)

    def delta(self, in_seconds=True):
        """Return time delta - empty (==0) if not tracked"""

        if self._tracked_time:
            delta = self.end_datetime - self.start_datetime
        else:
            delta = datetime.timedelta()

        if in_seconds:
            delta = delta.total_seconds()

        return delta


lib/ctt/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# 2012 Nico Schottelius (nico-ctt at schottelius.org)
#
# This file is part of ctt.
#
# ctt is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ctt is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ctt. If not, see <http://www.gnu.org/licenses/>.
#
#

import os
import os.path

VERSION = "1.1"
FILE_DELTA = "delta"
FILE_COMMENT = "comment"
DATEFORMAT = "%Y-%m-%d"
DATEFORMAT_PLAIN = DATEFORMAT.replace("%", "")
DATETIMEFORMAT = "%Y-%m-%d-%H%M"
DATETIMEFORMAT_PLAIN = DATETIMEFORMAT.replace("%", "")
REPORTFORMAT = "{start_datetime} ({delta}): {comment}"
AUTHOR = 'Nico Schottelius'
MAIL = 'nico-ctt at schottelius.org'
WWW = 'http://www.nico.schottelius.org/software/ctt/'
# Name of the folder to create - should not contain special characters
# to ensure cross-os compatibility
DISKFORMAT = DATETIMEFORMAT


class Error(Exception):
    pass


# Our output format
def user_timedelta(seconds):
    """Format timedelta for the user"""

    if seconds >= 3600:
        hours = int(seconds / 3600)
        seconds = seconds - (hours * 3600)
    else:
        hours = 0

    if seconds >= 60:
        minutes = int(seconds / 60)
        seconds = seconds - (minutes * 60)
    else:
        minutes = 0

    return (hours, minutes, seconds)


def ctt_dir():
    home = os.environ['HOME']
    ctt_dir = os.path.join(home, ".ctt")

    return ctt_dir


def project_dir(project):
    project_dir = os.path.join(ctt_dir(), project)

    return project_dir

    # return os.listdir(ctt_dir)


lib/ctt/listprojects.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# 2012 Nico Schottelius (nico-ctt at schottelius.org)
#
# This file is part of ctt.
#
# ctt is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ctt is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ctt. If not, see <http://www.gnu.org/licenses/>.
#
#

import ctt
import logging
import os

log = logging.getLogger(__name__)


class ListProjects(object):
    """Return existing projects"""

    @classmethod
    def commandline(cls, args):
        cls.print_projects()

    @classmethod
    def print_projects(cls):
        for project in cls.list_projects():
            print(project)

    @staticmethod
    def list_projects():
        for project in sorted(os.listdir(ctt.ctt_dir())):
            if project[0] == ".":
                continue

            yield project


completions/bash/ctt-completion.bash
_ctt()
{
    local cur prev opts cmds projects
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    opts="-h --help -d --debug -v --verbose"
    cmds="listprojects track report"

    case "${prev}" in
        track|report)
            if [[ -d ~/.ctt ]]; then
                projects=$(ls ~/.ctt)
            else
                projects=""
            fi
            ;;
    esac

    case "${prev}" in
        -*|listprojects)
            COMPREPLY=( $(compgen -W "${opts}" -- ${cur}) )
            return 0
            ;;
        track)
            opts="-h --help -d --debug -v --verbose --sd --start --ed --end -n --no-comment"
            COMPREPLY=( $(compgen -W "${opts} ${projects}" -- ${cur}) )
            return 0
            ;;
        report)
            opts="-h --help -d --debug -v --verbose --sd --start --ed --end -a --all -e --regexp -i --ignore-case -f -format -s --summary"
            COMPREPLY=( $(compgen -W "${opts} ${projects}" -- ${cur}) )
            return 0
            ;;
        *)
            ;;
    esac

    if [[ ${cur} == -* ]]; then 
    	COMPREPLY=( $(compgen -W "${opts}" -- ${cur}) )
    	return 0
    fi

    COMPREPLY=( $(compgen -W "${cmds}" -- ${cur}) )
    return 0
}

complete -F _ctt ctt


lib/ctt/report.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# 2012-2018 Nico Schottelius (nico-ctt at schottelius.org)
# 2016 Darko Poljak (darko.poljak at gmail.com)
#
# This file is part of ctt.
#
# ctt is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ctt is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ctt. If not, see <http://www.gnu.org/licenses/>.
#
#

import datetime

import logging

import os
import os.path
import re
import glob
import collections
import sys

import ctt
import ctt.listprojects

log = logging.getLogger(__name__)


class Report(object):
    """Create a report on tracked time"""

    def __init__(self, project, start_date, end_date,
                 output_format, regexp, ignore_case):

        self.project = project
        self.project_dir = ctt.project_dir(self.project)

        self.output_format = output_format
        self.regexp = regexp

        if ignore_case:
            self.search_flags = re.IGNORECASE
        else:
            self.search_flags = 0

        self._init_date(start_date, end_date)
        self._init_report_db()

    @classmethod
    def commandline(cls, args):
        # Report time for all projects
        if args.all:
            projects = ctt.listprojects.ListProjects.list_projects()

        else:
            projects = []
            for x in args.project:
                fnames = glob.glob(os.path.join(ctt.ctt_dir(), x))
                projects.extend(fnames)

        total_time = 0
        reports = collections.OrderedDict()
        for project in projects:
            report = cls(project=project, start_date=args.start,
                         end_date=args.end, output_format=args.output_format,
                         regexp=args.regexp, ignore_case=args.ignore_case)
            report_data = report.report()
            reports[report.project] = (report, report_data)
            total_time = total_time + report.total_time
        cls.print_reports(reports, args.output_format, args.summary)

        cls.summary(total_time)

    @staticmethod
    def print_report_time_entries(report_data, output_format, summary):
        ''' Print time entries from report_data report using output_format.
        '''
        keys = sorted(report_data.keys())
        for time in keys:
            entries = report_data[time]
            for entry in entries:
                print(output_format.format_map(entry))

    @staticmethod
    def print_reports(reports, output_format, summary):
        ''' Print reports using output_format for each entry.

            If summary is True then all time entries from all
            projects is extracted to one report dict.
            Otherwise, all time entries by each project is printed.
        '''
        if summary:
            summary_report = {}
        for project in reports:
            report, report_data = reports[project]
            if summary:
                for time in report_data:
                    if time not in summary_report:
                        summary_report[time] = report_data[time]
                    else:
                        summary_report[time].extend(report_data[time])
            else:
                report.header()
                Report.print_report_time_entries(report_data,
                                                 output_format, summary)
        # For summary do not print time entries.
        # if summary:
        #     Report.print_report_time_entries(summary_report,
        #             output_format, summary)

    def _init_date(self, start_date, end_date):
        """Setup date - either default or user given values"""

        now = datetime.datetime.now()
        first_day_this_month = now.replace(
            day=1, hour=0, minute=0, second=0, microsecond=0)
        next_month = first_day_this_month.replace(
            day=28) + datetime.timedelta(days=4)
        first_day_next_month = next_month.replace(day=1)
        last_day_this_month = first_day_next_month - datetime.timedelta(
            seconds=1)

        default_start_date = first_day_this_month
        default_end_date = last_day_this_month

        # default_end_date = first_day - datetime.timedelta(days=1)
        # default_start_date = default_end_date.replace(day=1)

        try:
            if start_date:
                self.start_date = datetime.datetime.strptime(
                    start_date[0], ctt.DATEFORMAT)
            else:
                self.start_date = default_start_date

            if end_date:
                self.end_date = datetime.datetime.strptime(
                    end_date[0], ctt.DATEFORMAT)
            else:
                self.end_date = default_end_date
        except ValueError as e:
            raise ctt.Error(e)

        self.end_date = self.end_date.replace(
            hour=23, minute=59, second=59)

        if self.start_date >= self.end_date:
            raise ctt.Error("End date must be after start date (%s >= %s)" %
                            (self.start_date, self.end_date))

    def _init_report_db(self):
        """Read all contents from db"""

        if not os.path.isdir(self.project_dir):
            raise ctt.Error("Project does not exist: %s" % (self.project))

        self._report_db = {}
        for dirname in os.listdir(self.project_dir):
            log.debug("Dirname: %s" % dirname)
            try:
                dir_datetime = datetime.datetime.strptime(
                    dirname, ctt.DISKFORMAT)
            except ValueError:
                log.warning("Invalid time entry {entry} for project "
                            "{project}, skipping.".format(
                                entry=dirname, project=self.project))
                continue

            if (dir_datetime >= self.start_date and
                    dir_datetime <= self.end_date):
                filename = os.path.join(
                    self.project_dir, dirname, ctt.FILE_DELTA)
                comment_filename = os.path.join(
                    self.project_dir, dirname, ctt.FILE_COMMENT)

                # Check for matching comment
                comment = None
                if os.path.exists(comment_filename):
                    with open(comment_filename, "r") as fd:
                        comment = fd.read().rstrip('\n')

                    # If regular expression given, but not matching, skip entry
                    if (self.regexp and
                            not re.search(self.regexp, comment,
                                          self.search_flags)):
                        continue

                self._report_db[dirname] = {}
                if comment:
                    self._report_db[dirname]['comment'] = comment

                with open(filename, "r") as fd:
                    self._report_db[dirname]['delta'] = fd.read().rstrip('\n')

                log.debug("Recording: %s: %s"
                          % (dirname, self._report_db[dirname]['delta']))

            else:
                log.debug("Skipping: %s" % dirname)

    def header(self):
        project_name = os.path.basename(self.project)
        print("Report for %s between %s and %s" %
              (project_name, self.start_date, self.end_date))

    @staticmethod
    def summary(total_time):
        hours, minutes, seconds = ctt.user_timedelta(total_time)

        print("Total time tracked: %sh %sm %ss." %
              (hours, minutes, seconds))

    @property
    def total_time(self):
        """Return all entries"""

        count = 0
        for entry in self._report_db.values():
            delta = entry['delta']
            log.debug("Adding %s to %s time..." % (delta, count))
            try:
                count = count + float(delta)
            except ValueError:
                log.warning("Invalid delta in entry {entry} for project "
                            "{project}, skipping for total time.".format(
                                entry=entry, project=self.project))
                continue
        return count

    def _get_report_entry(self, time, entry):
        ''' Get one time entry data.
        '''
        report = {}
        start_datetime = datetime.datetime.strptime(time, ctt.DATETIMEFORMAT)

        delta = datetime.timedelta(seconds=int(float(entry['delta'])))
        end_datetime = (start_datetime + delta).replace(microsecond=0)

        report['start_datetime'] = start_datetime.strftime(ctt.DATETIMEFORMAT)
        report['end_datetime'] = end_datetime.strftime(ctt.DATETIMEFORMAT)

        report['delta'] = delta
        report['delta_seconds'] = int(float(entry['delta']))
        report['delta_minutes'] = int(report['delta_seconds']/60)

        if 'comment' in entry:
            report['comment'] = entry['comment']
        else:
            report['comment'] = False
        return report

    def report(self):
        """Return total time tracked"""

        entries = {}
        time_keys = self._report_db.keys()
        for time in time_keys:
            entry = self._report_db[time]
            try:
                report = self._get_report_entry(time, entry)
            except ValueError:
                log.warning("Invalid delta in entry {entry} for project "
                            "{project}, skipping for report.".format(
                                entry=entry, project=self.project))
                continue
            if time not in entries:
                entries[time] = [report]
            else:
                entries[time].append(report)
        return entries
